### PR TITLE
Update Deprecations rules to latest RuboCop format

### PR DESCRIPTION
**Pull request by scottvidmar**
## Description
This refactors deprecations rules to use the latest format in use by RuboCop:
https://docs.rubocop.org/rubocop/development.html

No new functionality added.
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:
- [X ] I have read the **CONTRIBUTING** document.
- [ X] I have run the pre-merge tests locally and they pass.
- [ X] I have updated the documentation accordingly.
- [X ] I have added tests to cover my changes.
- [X ] All new and existing tests passed
- [ X] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

## Testing
Ran existing RSpec tests, as this is not adding or removing functionality.

$ bundle exec rspec spec -f p
Run options: include {:focus=>true}

All examples were filtered out; ignoring {:focus=>true}

Randomized with seed 304
................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

Finished in 0.88479 seconds (files took 2.08 seconds to load)
752 examples, 0 failures

Randomized with seed 304

### CODE REVIEW

File: lib/rubocop/cop/chef/deprecation/chef_handler_recipe.rb [hunk 1]
```diff
@@ -37,15 +37,12 @@ class ChefHandlerRecipe < Cop
             (send nil? :include_recipe (str {"chef_handler" "chef_handler::default"}))
           PATTERN
 
+          extend AutoCorrector
```
Comment (tas50) [inline lib/rubocop/cop/chef/deprecation/chef_handler_recipe.rb:40]: I'd prefer to have the extends right up with the class so it's clear that this class extends AutoCorrector

File: lib/rubocop/cop/chef/deprecation/chef_rewind.rb [hunk 2]
```diff
@@ -61,47 +61,54 @@ class ChefRewind < Cop
...
 
-          def autocorrect(node)
-            lambda do |corrector|
-              rewind_gem_install?(node) do
-                node = node.parent if node.parent&.block_type? # make sure we get the whole block not just the method in the block
-                corrector.remove(range_with_surrounding_space(range: node.loc.expression, side: :left))
-                return
-              end
+          def fixer(node, corrector)
+            rewind_gem_install?(node) do
```
Comment (tas50) [inline lib/rubocop/cop/chef/deprecation/chef_rewind.rb:94]: We should be able to extract this mess into each corrector not. It's one of the reasons they did the new autocorrect inline format.

File: lib/rubocop/cop/chef/deprecation/cheffile.rb [hunk 3]
```diff
@@ -22,18 +22,19 @@ module Chef
...
-        class Cheffile < Cop
+        class Cheffile < Base
           include RangeHelp
 
           MSG = 'The Librarian-Chef depsolving project is no longer maintained and a Cheffile should not be used for cookbook depsolving. Consider using Policyfiles instead.'
 
-          def investigate(processed_source)
-            return if processed_source.blank?
+          def on_new_investigation
+            f_name = processed_source.file_path
```
Comment (tas50) [inline lib/rubocop/cop/chef/deprecation/cheffile.rb:31]: We limit the filename in the config so there's no need to do the file name check again.

File: lib/rubocop/cop/chef/deprecation/chocolatey_package_uninstall_action.rb [hunk 4]
```diff
@@ -34,24 +34,21 @@ module ChefDeprecations
...
           MSG = 'Use the :remove action in the chocolatey_package resource instead of :uninstall which was removed in Chef Infra Client 14+'
 
+          extend AutoCorrector
           def on_block(node)
             match_property_in_resource?(:chocolatey_package, 'action', node) do |choco_action|
               choco_action.arguments.each do |action|
-                add_offense(action, location: :expression, message: MSG, severity: :warning) if action.source == ':uninstall'
+                add_offense(action, message: MSG, severity: :warning) do |corrector|
+                  corrector.replace(action, ':remove')
+                end if action.source == ':uninstall'
```
Comment (tas50) [inline lib/rubocop/cop/chef/deprecation/chocolatey_package_uninstall_action.rb:48]: It's a lot easier to follow these by removing the ```if``` on the block at line 48 and instead add ```next unless action.source == ':uninstall' ``` before the ```add_offense```

File: lib/rubocop/cop/chef/deprecation/deprecated_mixins.rb [hunk 5]
```diff
@@ -51,24 +51,29 @@ class UsesDeprecatedMixins < Cop
             (send nil? :require ( str {"chef/mixin/language" "chef/mixin/language_include_attribute" "chef/mixin/language_include_recipe"}))
           PATTERN
 
+          extend AutoCorrector
           def on_send(node)
             deprecated_mixin?(node) do
-              add_offense(node, location: :expression, message: MSG, severity: :warning)
+              add_offense(node, message: MSG, severity: :warning) do |corrector|
+                fixer(node, corrector)
```
Comment (tas50) [inline lib/rubocop/cop/chef/deprecation/deprecated_mixins.rb:58]: We should be able to call the individual autocorrects here as well

File: lib/rubocop/cop/chef/deprecation/chocolatey_package_uninstall_action.rb (hunk 4 above)
Comment (scottvidmar) [reply lib/rubocop/cop/chef/deprecation/chocolatey_package_uninstall_action.rb]: These should be addressed.

File: lib/rubocop/cop/chef/deprecation/deprecated_mixins.rb (hunk 5 above)
Comment (scottvidmar) [reply lib/rubocop/cop/chef/deprecation/deprecated_mixins.rb]: Should be all fixed up.

File: lib/rubocop/cop/chef/deprecation/cheffile.rb (hunk 3 above)
Comment (scottvidmar) [reply lib/rubocop/cop/chef/deprecation/cheffile.rb:31]: Fixed.

File: lib/rubocop/cop/chef/deprecation/chef_rewind.rb (hunk 2 above)
Comment (scottvidmar) [reply lib/rubocop/cop/chef/deprecation/chef_rewind.rb]: Fixed

File: lib/rubocop/cop/chef/deprecation/chef_handler_recipe.rb (hunk 1 above)
Comment (scottvidmar) [reply lib/rubocop/cop/chef/deprecation/chef_handler_recipe.rb]: Agreed. Fixed.

**Comment by scottvidmar** [comment]
Addressed PR feedback. Tests still passing:

 bundle exec rspec spec -f p
Run options: include {:focus=>true}

All examples were filtered out; ignoring {:focus=>true}

Randomized with seed 1869
................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

Finished in 1.3 seconds (files took 1.12 seconds to load)
752 examples, 0 failures

Randomized with seed 1869

File: lib/rubocop/cop/chef/deprecation/cheffile.rb (hunk 3 above)
Comment (tas50) [reply lib/rubocop/cop/chef/deprecation/cheffile.rb:31]: @scottvidmar This doesn't look fixed yet

File: lib/rubocop/cop/chef/deprecation/launchd_deprecated_hash_property.rb [hunk 6]
```diff
@@ -33,23 +33,20 @@ module ChefDeprecations
...
           def on_block(node)
-            match_property_in_resource?(:launchd, 'hash', node) do |hash_prop|
-              add_offense(hash_prop, location: :expression, message: MSG, severity: :warning)
-            end
-          end
-
-          def autocorrect(node)
-            lambda do |corrector|
-              corrector.replace(node.loc.expression, node.loc.expression.source.gsub(/^hash/, 'plist_hash'))
+            return unless match_property_in_resource?(:launchd, 'hash', node) do |offense|
```
Comment (tas50) [inline lib/rubocop/cop/chef/deprecation/launchd_deprecated_hash_property.rb:46]: There shouldn't be a need for the return unless bit here. If it matches it'll run the block

Comment (scottvidmar) [reply lib/rubocop/cop/chef/deprecation/launchd_deprecated_hash_property.rb]: Good call. Removed that bit.

File: lib/rubocop/cop/chef/deprecation/cheffile.rb (hunk 3 above)
Comment (scottvidmar) [reply lib/rubocop/cop/chef/deprecation/cheffile.rb]: Now it should be fixed.

**Comment by tas50** [comment]
Thanks for taking the time to fix these all up. It'll allow us to upgrade RuboCop whenever they rip out the old classes.

**Pull request merged**